### PR TITLE
Send org key in api requests cat v1

### PR DESCRIPTION
**Pull request by nanditagoclimate**
The APIs will now accept a param called 'key' where we can pass the org key.
Refer to API doc - https://docs.google.com/document/d/1Yxu9UyQzFU_vUlBI1W-cWzvHPZCVw4oJ12BprC835-A/edit#heading=h.pg3ahgw5lf9a

We fetch organization details from the param key and send data if any of the below conditions is met:
1. logged in user is a super admin

<or>

2.logged in user is an org admin for that organization

While creating an organization through Admin tool, key is now made mandatory.

### CODE REVIEW

File: app/controllers/api/v1/api_controller.rb [hunk 1]
```diff
@@ -15,6 +15,10 @@ def url_options
         super
       end
 
+      def set_org_for_super_admin(org_key)
```
Comment (kallenilver) [inline app/controllers/api/v1/api_controller.rb:18]: Seems like change of names here could be a good idea if the method only finds an org and does not set an org?

Comment (kallenilver) [reply app/controllers/api/v1/api_controller.rb:18]: Or should it set something that is missing?

**Review by kallenilver — APPROVED** [review]

**Review by axelsson — COMMENTED** [review]
Since the frontend doesn't know or care for whether the user is a super admin or not, there's no need to handle "no key sent", there will always be an org key included so I think we should always find the organization by that key and ignore the org that the user is connected to.

I guess we will want the users to be org_admins as CAT will not only be read only mode and we've lot's of users connected to organizations. Seeing CAT in the menu requires admin role atm (as far as I remember).

We should probably also auth the users when they edit climate actions for businesses.

File: app/controllers/api/v1/business_climate_actions_controller.rb [hunk 2]
```diff
@@ -6,9 +6,13 @@ class BusinessClimateActionsController < ApiController
       before_action :authenticate_user!
 
       def show
-        organization_id = current_user.organization&.id
+       return unless params['key']
```
Comment (axelsson) [inline app/controllers/api/v1/business_climate_actions_controller.rb:9]: indentation is off here

File: app/models/organization.rb [hunk 3]
```diff
@@ -6,7 +6,9 @@ class Organization < ApplicationRecord
   has_many :users
   has_many :user_admins
 
-  validates_presence_of :name
+  validates_presence_of :name, :key
+  validates_uniqueness_of :name, :key
```
Comment (axelsson) [inline app/models/organization.rb:10]: This prevents creation of a new org in admin without a key, is that intentional?

**Review by axelsson — COMMENTED** [review]
Lots of code duplication, here since all affected endpoints require the org and the check for access. If it's possible to break it out into a common function somewhere that would be nice. Not sure what context would be the most suitable for that though?

File: app/controllers/api/v1/business_climate_actions_controller.rb (hunk 2 above)
Comment (axelsson) [reply app/controllers/api/v1/business_climate_actions_controller.rb:9]: Also I find it strange that the API returns http 204 if the key is missing or if the key belongs to a user that doesn't have access, should probably be 4XX (not sure?) and 403 for those.

File: app/models/organization.rb (hunk 3 above)
Comment (nanditagoclimate) [reply app/models/organization.rb:10]: Yes.

**Comment by nanditagoclimate** [comment]
I see your point. We have something called helpers but that's where common business logic goes but since these are just checks they need to be at controller level itself. This is also more readable. Further refactoring would be over-engineering.

**Review by axelsson — APPROVED** [review]
I tried this out with CAT, works smooth 👍

**Pull request merged**